### PR TITLE
Closes #505: enhance is_valid_dtc() [create website]

### LINKED ISSUE #505: Feature Request:  `is_valid_dtc` check the number value too
### Feature Idea

the current `is_valid_dtc()` only check the format,  can it also do some check for number value?

https://github.com/pharmaverse/admiraldev/blob/9233639da0c65c0e80a756203bc1e5aa7c337da2/R/is.R#L48-L60

like the below AI assistant code 

### Relevant Input

```r
DTC <- c(
  "2025-13",                 # Month 13
  "2025-08-32",              # Day 32
  "2025-08-10T24:30:45",     # Hour 24
  "2025-08-10T14:60:45",     # Minute 60
  "2025-08-10T14:30:60"      # Second 60
)
admiraldev::is_valid_dtc(DTC)
```

### Relevant Output

```
#> [1] TRUE TRUE TRUE TRUE TRUE
```

### Reproducible Example/Pseudo Code

``` r
DTC <- c(
  "2025-13",                 # Month 13
  "2025-08-32",              # Day 32
  "2025-08-10T24:30:45",     # Hour 24
  "2025-08-10T14:60:45",     # Minute 60
  "2025-08-10T14:30:60"      # Second 60
)

admiraldev::is_valid_dtc(DTC)
#> [1] TRUE TRUE TRUE TRUE TRUE
```

The code with AI assisted

``` r
is_valid_dtc <- function(dtc) {
  
  if (length(dtc) == 1) {
    if (is.na(dtc)) {
      return(TRUE)
    }
  }
 
  # Define regex components for validation and extraction
  year_pattern <- "(?:\\d{4}|-)"                           # Year: 4 digits OR single dash
  month_pattern <- "(?:0[1-9]|1[0-2]|-)"                   # Month: 01-12 OR single dash
  day_pattern <- "(?:0[1-9]|[12]\\d|3[01]|-)"              # Day: 01-31 OR single dash
  hour_pattern <- "(?:[01]\\d|2[0-3]|-)"                   # Hour: 00-23 OR single dash
  minute_pattern <- "(?:[0-5]\\d|-)"                       # Minute: 00-59 OR single dash
  second_pattern <- "(?:[0-5]\\d|-)"                       # Second: 00-59 OR single dash
  fraction_pattern <- "(?:\\.\\d{1,5})?"                   # Optional fraction: .1 to .12345
  
  # Build comprehensive validation pattern step by step
  dtc_pattern <- paste0(
    "^$",                                                   # Allow empty string
    "|",                                                    # OR
    "^", year_pattern,                                      # Required: Year (4 digits or -)
    "(?:",                                                  # Optional group: Month and following
    "-", month_pattern,                                     # Month with preceding dash
    "(?:",                                                  # Optional group: Day and following
    "-", day_pattern,                                       # Day with preceding dash
    "(?:",                                                  # Optional group: Time components
    "T", hour_pattern,                                      # Hour with preceding T
    "(?:",                                                  # Optional group: Minutes and following
    ":", minute_pattern,                                    # Minute with preceding colon
    "(?:",                                                  # Optional group: Seconds
    ":", second_pattern,                                    # Second with preceding colon
    fraction_pattern,                                       # Optional fractional seconds
    ")?",                                                   # End seconds group
    ")?",                                                   # End minutes group
    ")?",                                                   # End time group
    ")?",                                                   # End day group
    ")?",                                                   # End month group
    "$"                                                     # End of string
  )
  
  return(stringr::str_detect(dtc, dtc_pattern) | is.na(dtc))
}

test_valid_cases <- c(
  # === VALID CASES ===
  
  # Empty string
  "",
  
  # Complete valid formats
  "2025-08-10T14:30:45.123",              # Full with fraction
  "2025-08-10T14:30:45",                  # Full without fraction
  
  # Missing parts at the end (truncated)
  "2025-08-10T14:30",                     # Missing second
  "2025-08-10T14",                        # Missing minute & second
  "2025-08-10",                           # Missing time
  "2025-08",                              # Missing day & time
  "2025",                                 # Missing month, day & time
  
  # Missing parts in the middle
  "2025---10T14:30:45",                   # Missing month
  "2025-08--T14:30:45",                   # Missing day
  "2025-08-10T-:30:45",                   # Missing hour
  "2025-08-10T14:-:45",                   # Missing minute
  "2025----T14:30:45",                    # Missing month & day
  "2025-08-10T-:-:45",                    # Missing hour & minute
  "2025---10T-:30:45",                    # Missing month & hour
  "2025-08--T-:30:45",                    # Missing day & hour
  "2025-08--T14:-:45",                    # Missing day & minute
  "2025----T-:30:45",                     # Missing month, day & hour
  "2025----T-:-:45",                      # Missing month, day, hour & minute
  
  # Missing parts at the beginning
  "--08-10T14:30:45",                     # Missing year
  "----10T14:30:45",                      # Missing year & month
  "-----T14:30:45",                       # Missing year, month & day
  "-----T-:30:45",                        # Missing year, month, day & hour
  "-----T-:-:45",                         # Missing year, month, day, hour & minute
  
  # Missing combinations
  "--08-10T14:30",                        # Missing year & second
  "--08-10T14:-:45",                      # Missing year & minute
  "--08-10T-:30:45",                      # Missing year & hour
  "--08--T14:30:45",                      # Missing year & day
  "----10T-:30:45",                       # Missing year, month & hour
  "----10T14:-:45",                       # Missing year, month & minute
  "-----T14:30:45",                       # Missing date
  "-----T-:30:45",                        # Missing date & hour
  "-----T14:-:45",                        # Missing date & minute
  "-----T-:-:45",                         # Missing date, hour & minute
  "--08-10",                              # Missing year & time
  "2025---10",                            # Missing month & time
  "----10",                               # Missing year, month & time
  "--08",                                 # Missing year, day & time
  "----10T14",                            # Missing year, month & minute & second
  
  # Valid edge cases
  "0000-01-01T00:00:00",                  # Minimum values
  "9999-12-31T23:59:59",                  # Maximum values
  "2025-08-10T14:30:45.1",                # Single digit fraction
  "2025-08-10T14:30:45.12345"             # Maximum fraction digits
)

all(is_valid_dtc(test_valid_cases))
#> [1] TRUE


test_invalid_cases <- c(
  # === INVALID CASES ===
  
  # Invalid year formats
  "25",                                   # Year too short
  "12345",                                # Year too long
  "abcd",                                 # Non-numeric year
  
  # Invalid month values
  "2025-00",                              # Month 00
  "2025-13",                              # Month 13
  "2025-1",                               # Single digit month
  "2025-ab",                              # Non-numeric month
  
  # Invalid day values
  "2025-08-00",                           # Day 00
  "2025-08-32",                           # Day 32
  "2025-08-1",                            # Single digit day
  "2025-08-ab",                           # Non-numeric day
  
  # Invalid hour values
  "2025-08-10T24:30:45",                  # Hour 24
  "2025-08-10T1:30:45",                   # Single digit hour
  "2025-08-10Tab:30:45",                  # Non-numeric hour
  
  # Invalid minute values
  "2025-08-10T14:60:45",                  # Minute 60
  "2025-08-10T14:1:45",                   # Single digit minute
  "2025-08-10T14:ab:45",                  # Non-numeric minute
  
  # Invalid second values
  "2025-08-10T14:30:60",                  # Second 60
  "2025-08-10T14:30:1",                   # Single digit second
  "2025-08-10T14:30:ab",                 # Non-numeric second
  
  # Trailing colon cases (reported)
  "2025-08-10T14:",                        # Colon without minute (should be invalid)
  "2025-08-10T14:30:",                     # Colon without second (should be invalid)
  
  # Invalid separators
  "2025/08/10T14:30:45",                  # Wrong date separator
  "2025-08-10 14:30:45",                  # Space instead of T
  "2025-08-10T14-30-45",                  # Wrong time separator
  
  # Incomplete formats
  "2025-",                                # Trailing dash after year
  "2025-08-",                             # Trailing dash after month
  "2025-08-10T",                          # T without time
  "2025-08-10T14:",                       # Colon without minute
  "2025-08-10T14:30:",                    # Colon without second
  
  # Wrong structure
  "12-25",                                # Missing year indicator (should be --12-25)
  "T14:30:45",                            # Time without date (should be -----T14:30:45)
  "2025-08-T14:30:45",                    # Missing day with just dash
  "2025--T14:30:45",                      # Missing month and day with wrong format
  
  # Invalid fraction
  "2025-08-10T14:30:45.",                 # Fraction without digits
  "2025-08-10T14:30:45.123456",           # Too many fraction digits
  
  # Multiple separators
  "2025--08-10T14:30:45",                 # Double dash in month
  "2025-08--10T14:30:45",                 # Double dash in day
  "2025-08-10TT14:30:45",                 # Double T
  "2025-08-10T14::30:45",                 # Double colon
  
  # Empty components
  "2025--08-10T14:30:45",                 # Empty month position
  "2025-08-10T:30:45"                     # Empty hour position
)

all(!is_valid_dtc(test_invalid_cases))
#> [1] TRUE
```

### PR DESCRIPTION
**Pull request by manciniedoardo**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/cran-release/CONTRIBUTING.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiraldev/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

To create the website for the content of this PR add "[create website]" to the title of the pull request.

### CODE REVIEW

**Comment by manciniedoardo** [comment]
@adcascone @bundfussr this is now ready to review.

File: NEWS.md [hunk 1]
```diff
@@ -6,6 +6,8 @@
 
 ## Updates of Existing Functions
 
+- `is_valid_dtc()` was enhanced with the arguments `check_dtc` to allow users
+to check whether the input date/datetime is a real date/datetime`. (#505)
```
Comment (bundfussr) [inline NEWS.md]:
```suggestion
to check whether the input date/datetime is a real date/datetime. (#505)
```

File: R/is.R [hunk 2]
```diff
@@ -36,16 +36,30 @@ is_order_vars <- function(arg) {
...
+#'
+#' @details If `check_dtc` is `FALSE`, the function only checks if the format of the
+#' string is valid, i.e. it will not check if the actual date/datetime is real, so
+#' calls such as `is_valid_dtc("2020-13", check_dtc = FALSE))` and
+#' `is_valid_dtc("2020-12-01T25:00:00", check_dtc = FALSE)` will return`TRUE`.
+#'
 #' @export
 #' @keywords is
 #' @family is
 #'
```
Comment (bundfussr) [inline R/is.R:61]: Could you add some examples?

File: R/is.R [hunk 3]
```diff
@@ -36,16 +36,30 @@ is_order_vars <- function(arg) {
...
+#' date or datetime (e.g. "2020-02-31" is not real). Note that `NA` is a valid
+#' datetime.
+#'
+#' @permitted [boolean]
 #'
 #' @return `TRUE` if the argument is a valid `--DTC` string, `FALSE` otherwise
+#'
+#' @details If `check_dtc` is `FALSE`, the function only checks if the format of the
+#' string is valid, i.e. it will not check if the actual date/datetime is real, so
+#' calls such as `is_valid_dtc("2020-13", check_dtc = FALSE))` and
```
Comment (bundfussr) [inline R/is.R]:
```suggestion
#' calls such as `is_valid_dtc("2020-13", check_dtc = FALSE)` and
```

File: tests/testthat/test-is.R [hunk 4]
```diff
@@ -20,13 +20,20 @@ test_that("is_valid_dtc Test 3: returns TRUE if input are valid dtc", {
 })
 
 # is_valid_dtc ----
-## Test 4: returns error if input if input are NOT valid dtc  ----
-test_that("is_valid_dtc Test 4: returns error if input if input are NOT valid dtc ", {
+## Test 4: returns error if input are NOT valid dtc  ----
+test_that("is_valid_dtc Test 4: returns error if input are NOT valid dtc ", {
```
Comment (bundfussr) [inline tests/testthat/test-is.R]:
```suggestion
test_that("is_valid_dtc Test 4: returns FALSE if input are NOT valid dtc ", {
```

File: tests/testthat/test-is.R [hunk 5]
```diff
@@ -20,13 +20,20 @@ test_that("is_valid_dtc Test 3: returns TRUE if input are valid dtc", {
...
   expect_false(is_valid_dtc("2021-03-T01:20:30"))
 })
 
+## Test 5: returns error if input date or datetime is not real date  ----
+test_that("is_valid_dtc Test 5: returns error if input date or datetime is not real date", {
```
Comment (bundfussr) [inline tests/testthat/test-is.R]:
```suggestion
test_that("is_valid_dtc Test 5: returns FALSE if input date or datetime is not real date", {
```

File: tests/testthat/test-is.R [hunk 6]
```diff
@@ -20,13 +20,20 @@ test_that("is_valid_dtc Test 3: returns TRUE if input are valid dtc", {
 })
 
 # is_valid_dtc ----
```
Comment (bundfussr) [inline tests/testthat/test-is.R]:
```suggestion
```

File: tests/testthat/test-is.R [hunk 7]
```diff
@@ -20,13 +20,20 @@ test_that("is_valid_dtc Test 3: returns TRUE if input are valid dtc", {
...
   expect_false(is_valid_dtc("2021-03-T01:20:30"))
 })
 
+## Test 5: returns error if input date or datetime is not real date  ----
+test_that("is_valid_dtc Test 5: returns error if input date or datetime is not real date", {
+  expect_false(is_valid_dtc("2021-02-31", check_dtc = TRUE))
+  expect_false(is_valid_dtc("2021-02-29", check_dtc = TRUE)) # not a leap year!
+  expect_false(is_valid_dtc("2021-02-28T25:00:00", check_dtc = TRUE))
+})
+
```
Comment (bundfussr) [inline tests/testthat/test-is.R:33]: Could you add a test for `check_dtc = TRUE` which returns `TRUE`?

File: R/is.R (hunk 2 above)
Comment (manciniedoardo) [reply R/is.R:61]: done

File: tests/testthat/test-is.R (hunk 7 above)
Comment (manciniedoardo) [reply tests/testthat/test-is.R:33]: done

File: R/is.R [hunk 8]
```diff
@@ -36,16 +36,45 @@ is_order_vars <- function(arg) {
...
+#' ## Format is valid, date/datetimes are real
+#' is_valid_dtc("2020-02")
+#' is_valid_dtc("2020-02-28")
+#' is_valid_dtc("2020-02-28T14:43")
+#'
+#' # Format is valid, date/datetimes are not real but this is not checked
+#' is_valid_dtc("2020-02-31")
+#' is_valid_dtc("2020-02-28T25:00:00")
+#'
+#' # Format is valid, date/datetimes are not real and this is not checked
```
Comment (bundfussr) [inline R/is.R]:
```suggestion
#' # Format is valid, date/datetimes are not real and this is checked
```

**Review by bundfussr — APPROVED** [review]

**Comment by manciniedoardo** [comment]
@adcascone if you don't mind, I'll go ahead and merge this one as I need it for https://github.com/pharmaverse/admiral/pull/3134 - if you still want to take a look feel free to leave some comments and we can make another PR. sorry for the rush... time crunch!
